cps/kni.c

```diff
@@ -633,16 +688,24 @@ rd_modroute(const struct nlmsghdr *req, void *arg)
 {
 	struct nlattr *tb[__RTA_MAX] = {};
 	struct rtmsg *rm = mnl_nlmsg_get_payload(req);
-	struct route_update *update = arg;
-
-	update->valid = false;
+	struct route_updates *updates = arg;
+	struct route_update *update;
 
 	CPS_LOG(DEBUG, "cps update: [%s] family=%u dst_len=%u src_len=%u tos=%u table=%u protocol=%u scope=%u type=%u flags=%x\n",
 		req->nlmsg_type == RTM_NEWROUTE ? "NEW" : "DEL",
 		rm->rtm_family, rm->rtm_dst_len, rm->rtm_src_len,
 		rm->rtm_tos, rm->rtm_table, rm->rtm_protocol,
 		rm->rtm_protocol, rm->rtm_scope, rm->rtm_flags);
 
+	if (updates->count == MAX_RTNL_UPDATES) {
+		CPS_LOG(ERR, "Too many routing table updates (more than %lu) in one Netlink packet; ignoring this update\n",
+			MAX_RTNL_UPDATES);
+		return MNL_CB_OK;
```

Comment:
That error code in libmnl in this context just means to stop processing any more Netlink messages in the current packet, which we could do. The way it is now, any other Netlink messages in the packet will continue to be processed and logged (if CPS logging is set to debug), but won't be acted upon.

Reply:
I understand that `MNL_CB_ERROR` stops processing the packet AND returns an error. It's also my understanding that accepting the whole packet and not returning the acknowledgments or errors to each request will make the daemon hang. On top of it, if `count
updates->count` goes beyond its maximum, `MAX_RTNL_UPDATES` is not correct.

Finally, let's make the test robust replacing the `==` with `>=`.

Reply:
Returning `MNL_CB_ERROR` doesn't send any error message back to sender, we have to send it ourselves (which we do in the next patch). Would it be better to reorder these patches so that we send acknowledgements/errors first and then have this patch?

If not, yes I can replace the test to make it more robust.

Reply:
We don't need to change the order of the patches. Adding a TODO to address in the second patch is fine. In any case, we need to report the error the daemon.

Curiosity, is `MNL_CB_STOP` equal to  `MNL_CB_ERROR`?

Reply:
No, they are different. `MNL_CB_STOP` is used to indicate that the callback runqueue should be stopped for some non-error, and `MNL_CB_ERROR` means it should be stopped due to an error.